README.md

```diff
@@ -0,0 +1,812 @@
+
+Initializes the Opus context with the specified sample rate.
+
+A result code is returned.
+
+#### reset()
+
+Resets the Opus context state.
+
+A result code is returned.
+
+#### channels()
+
+Returns the number of channels the context was initialized for.
+
+#### sampleRate()
+
+Returns the sample rate the context was initialized for.
+
+#### inDTX()
+
+Returns whether the context is in DTX mode.
+
+It is when last frame was either ignored or a comfort noise frame.
+
+#### phaseInversion()
+
+Returns whether phase inversion is enabled.
+
+#### togglePhaseInversion()
+
+Enables or disables phase inversion, depending on the passed flag.
+
+Returns whether the operation succeeded.
+
+#### packetSamples()
+
+Returns the number of samples in the specified Opus packet.
+
+### OpusEncoder
+
+Used to encode Opus packets.
+
+An object can be constructed through:
+
+- Move
+- Number of channels
+
+#### operator()
+
+Encodes the specified data into the specified buffer.
+
+#### init()
+
+Initializes the Opus context with the specified sample rate and preset.
+
+A result code is returned.
+
+#### reset()
+
+Resets the Opus context state.
+
+A result code is returned.
+
+#### channels()
+
+Returns the number of channels the context was initialized for.
+
+#### sampleRate()
+
+Returns the sample rate the context was initialized for.
+
+#### preset()
+
+Returns the context's current preset.
+
+#### setPreset()
+
+Sets the specified preset in the context.
+
+Returns whether the operation succeeded.
+
+#### bitrate()
+
+Returns the context's current bitrate.
+
+#### setBitrate()
+
+Sets the specified bitrate in the context.
+
+Special values are:
+
+- `0` to set the default value.
+- `infinite32` to set the maximum value supported.
+
+Returns whether the operation succeeded.
+
+#### inDTX()
+
+Returns whether the context is in DTX mode.
+
+It is when last frame was either ignored or a comfort noise frame.
+
+#### phaseInversion()
+
+Returns whether phase inversion is enabled.
+
+#### togglePhaseInversion()
+
+Enables or disables phase inversion, depending on the passed flag.
+
+Returns whether the operation succeeded.
+
+#### vbr()
+
+Returns whether variable bitrate is enabled.
+
+#### toggleVBR()
+
+Enables or disables variable bitrate, depending on the passed flag.
+
+Returns whether the operation succeeded.
+
+### Pack
+
+Used to serialize and deserialize Protobuf messages.
+
+Its internal buffer is a vector of bytes and can hold a serialized message, ready to be sent.  
+The first 6 bytes of the buffer represent the header as a `NetHeader` object.
+
+An object can be constructed through:
+
+- Copy
+- Move
+- `Message` (native or Protobuf's)
+- `NetHeader` (+ data size if UDP)
+
+Allocating extra space for the internal buffer is possible and it's what `Peer` does internally for UDP, to improve efficiency.  
+The buffer size is set to 1024 bytes, which is the formal maximum for our protocol, and the `Pack` object is recycled for every packet received.
```

Comment:
We should drop this formal constraint from our protocol and thus also not use it in libmumble.

Reply:
No problem, we can just pass the desired buffer size when starting the UDP engine.

Reply:
d5215ef0a3f8162ad816d082086a57744fbc1683

Reply:
Wel that's not really what I meant. With this the package size still has a hard limit depending on what the library user chose to use. Instead we should adapt the size to however large the incoming packet is (with some sanity checking to make sure maliciously crafted packages can drown the peer)

Reply:
In terms of efficiency it's much better to know in advance the maximum reasonable packet size.

It's fairly straightforward to calculate that, since we assume only Opus is used.

Reply:
I know, but having a hard-limit programmed into the application makes future extensions of the protocol difficult as older clients will outright reject any packet that exceeds what they consider the maximum package size.

And if you reuse the buffer for the packages, I think you get about the same efficiency as you essentially only need to resize the buffer once or twice perhaps and then its size will remain constant.

Reply:
> I know, but having a hard-limit programmed into the application makes future extensions of the protocol difficult as older clients will outright reject any packet that exceeds what they consider the maximum package size.

Fair point. Could you open an issue for it, please?

> And if you reuse the buffer for the packages, I think you get about the same efficiency as you essentially only need to resize the buffer once or twice perhaps and then its size will remain constant.

Yes, of course.